### ISSUE TITLE
Hanging forever after receiving results

### ISSUE THREAD

**Issue by fmaule** [swellstores/swell-node#10]
I am trying to run a simple CLI script that returns an order using `swell-node`, however for some reason it gets stuck and keeps hanging forever.

Example POC:

```
const swell = require('swell-node');
swell.init(SWELL_SHOP, SWELL_KEY);

const main = async () => {
  const orders = await swell.get('/orders', { id: 'some-swell-id' })
  const { count, results } = orders
  console.log(count)
}

main()
```

Response is being received successfully and then the process gets stuck forever.
Things I've tried: enabling/disabling cache, decreasing the timeout, adding DEBUG=* (no debug available), switching from node14 to node10, tracing with `why-is-node-running`: 

```
# TCPWRAP
some-path/node_modules/swell-node/lib/connection.js:41 - this.stream = proto.connect(
some-path/node_modules/swell-node/lib/connection.js:33 - this.connect(callback);
some-path/node_modules/swell-node/lib/client.js:70     - this.server = new Connection(
some-path/node_modules/swell-node/lib/client.js:124    - this.connect();
some-path/node_modules/swell-node/lib/client.js:276    - return this.request('get', url, data, callback);

# TLSWRAP
some-path/node_modules/swell-node/lib/connection.js:41 - this.stream = proto.connect(
some-path/node_modules/swell-node/lib/connection.js:33 - this.connect(callback);
some-path/node_modules/swell-node/lib/client.js:70     - this.server = new Connection(
some-path/node_modules/swell-node/lib/client.js:124    - this.connect();
some-path/node_modules/swell-node/lib/client.js:276    - return this.request('get', url, data, callback);

# DNSCHANNEL
(unknown stack trace)
```

Currently out of ideas. Any suggestions? Thanks!

**Comment by fmaule** [comment]
Digging a little bit more it appears that the `$end` response is never being received, hence the `close` event is never being emitted.

This is the only response i'm getting:

```
{
  '$links': {
    cart: { url: '/carts/{cart_id}' },
    items: { links: [Object] },
    billing: { links: [Object] },
    shipping: { links: [Object] },
    coupon: { url: '/coupons/{coupon_id}' },
    giftcards: { links: [Object] },
    promotions: { url: '/promotions/{promotion_ids}' },
    authorized_payment: { url: '/payments/{authorized_payment_id}' },
    parent: { url: '/orders/{parent_id}' },
    prev: { url: '/orders/{prev_id}' },
    next: { url: '/orders/{next_id}' },
    account: { url: '/accounts/{account_id}' },
    subscription: { url: '/subscriptions/{subscription_id}' },
    shipments: { url: '/shipments?order_id={id}' },
    returns: { url: '/returns?order_id={id}' },
    invoices: { url: '/invoices?order_id={id}' },
    pending_invoices: { url: '/invoices?order_id={id}&status=pending' },
    credits: { url: '/credits?order_id={id}' },
    pending_credits: { url: '/credits?order_id={id}&status=pending' },
    payments: { url: '/payments?order_id={id}' },
    refunds: { url: '/payments:refunds?order_id={id}' },
    notifications: { url: '/notifications?record_id={id}' }
  },
  '$url': '/orders',
  '$collection': 'com.orders',
  '$data': { count: 0, results: [], page: 1 },
  '$time': 28,
  '$status': 200,
  '$cached': { 'com.orders': 5510 }
}
```


https://github.com/swellstores/swell-node/blob/bc599a6f2415f1ff7fb2703fae69a8ee132b8e91/lib/connection.js#L175

Forcing it with a `if (response.$end || response.$status === 200)` actually fixes the issue, but I believe this is custom protocol I have no clue about ☹️

**Comment by greph** [comment]
Hey @fmaule , thanks for opening the issue. I've forwarded the request onto our platform team to take a look.

**Comment by fmaule** [comment]
Hey @greph, thanks, any updates on this?

**Comment by greph** [comment]
Hey @fmaule , no update yet, though I've bumped the issue with our engineers. Thanks for your patience.

**Comment by bettiolo** [comment]
Are there any updates on this? We are having the same problem!

**Comment by greph** [comment]
Hey @fmaule, @bettiolo, apologies for the delayed response. The close() method can be used to terminate the script, otherwise the connection is designed to remain open indefinitely.

```
const swell = require('swell-node');
swell.init(SWELL_SHOP, SWELL_KEY);

const main = async () => {
  ...
   swell.close()
  ...
}

main()
```

**Comment by fmaule** [comment]
Hey @greph, thanks, that was indeed the first thing I've tried and it doesn't work.

As said, the stream is terminated only if we receive an `$end` in the response - that is not happening and is on your side (custom wire protocol). I cannot really do anything, other than proxying the request and forcing and $end or forking the module to fix it. 
I'd rather avoid both options, honestly, and probably should just stick with the standard HTTP API not using this given it's not working?

**Comment by fmaule** [comment]
@greph this issue has been open for a while now, at this point I'm considering it as a `WONTFIX` and will use the HTTP API for the time being.
Thanks for your help anyway

**Comment by greph** [comment]
Hi @fmaule , appreciate your patience on the issue. I've raised it with our engineering team to see if there are any other workaround available.

**Comment by greph** [comment]
@fmaule  if possible, can you provide a code snippet where you are manually calling the close method? We're attempting to reproduce the issue.

**Comment by fmaule** [comment]
@greph hey, sorry for the late reply. 
Using `swell.close()` actually works, it still takes a bit (~2/3 seconds) but eventually ends the process.
Quite sure that's the first thing we've tried but probably got tricked by the long shutdown time.

Thanks!

**Comment by woutercouvaras** [comment]
We're experiencing the same issue and waiting on support for further feedback, but given that this has existed since 2021, I'm guessing support won't come back with anything useful.  

I've just installed the `swell-node-http` package, which is just a wrapper around axios.  It's pretty much a like-for-like api, so your code will likely not need to change.  It is noticeably slower though, but I'd rather havce slow and working than fast and unreliable.

I'm still waiting on the support team to confirm if its safe to use this package, but right now, that's our only choice.  If we can't use it, we'll have to move to another system because `swell-node` is just not viable in it's current form.
